torch/distributions/transforms.py

```diff
@@ -0,0 +1,290 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all
+from torch.nn.functional import sigmoid
+
+__all__ = [
+    'AbsTransform',
+    'AffineTransform',
+    'BoltzmannTransform',
+    'CachedTransform',
+    'ExpTransform',
+    'InverseTransform',
+    'SigmoidTransform',
+    'StickBreakingTransform',
+    'Transform',
+]
+
+
+class Transform(object):
+    """
+    Abstract class for transformations with computable inverse log
+    det jacobians. They are primarily used in
+    :class:`torch.distributions.TransformedDistribution`.
+
+    Derived classes should implement one or both of :meth:`_forward` or
+    :meth:`_inverse` and should implement :meth:`log_abs_det_jacobian`.
+    Derived classes may store intermediate results in the `._cache` dict.
+    """
+    bijective = False
+
+    def __eq__(self, other):
+        return type(other) is type(self)
+
+    def __ne__(self, other):
+        # Necessary for Python2
+        return not self.__eq__(other)
+
+    def forward(self, x):
+        """
+        Abstract method to compute forward transformation.
+        """
+        raise NotImplementedError
+
+    def inverse(self, y):
+        """
+        Abstract method to compute inverse transformation.
+        """
+        raise NotImplementedError
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Computes the log det jacobian `log |dy/dx|` given input and output.
+        """
+        raise NotImplementedError
+
+
+class CachedTransform(Transform):
+    """
+    Abstract base class for transforms that implement one of :meth:`forward`
+    or :meth:`backward` via by caching the latest value (i.e. an LRU(1) cache).
+
+    This class is useful for tranforms whose inverses are either expensive or
+    numerically unstable. Note that care must be taken with memoized values
+    since the autograd graph may be reversed. For example while the following
+    works::
+
+        y = t.forward(x)
+        t.log_abs_det_jacobian(x, y).backward()  # x will receive gradients.
+
+    However the following will error due to dependency reversal::
+
+        y = t.forward(x)
+        z = t.inverse(y)
+        grad(z.sum(), [y])  # error because z is x
+
+    Derived classes should implement one or both of :meth:`_forward` and
+    :meth:`_backward`.
+    """
+    def __init__(self):
+        self._cached_x_y = None, None
+
+    def forward(self, x):
+        """
+        Invokes the memoized transform `x => y`.
+        """
+        x_old, y_old = self._cached_x_y
+        if x is x_old:

```

Comment:
@apaszke Is it possible in Python to the version counter of `x` to avoid subtle errors on in-place modification? E.g. could store
```py
self._cached_x_y = x, x.version, y, y.version
```

Reply:
Well, if we decided we wanted to, it would be a simple matter to expose it to Python. Not sure if we want to though.

Reply:
It fine to ignore `.version`. I just wanted to point out this possible use case.

Reply:
I think we should verify the version. Otherwise the code can silently give incorrect results, which we're trying to avoid.